http-clients/src/main/java/org/triplea/http/client/web/socket/WebsocketListener.java

```diff
@@ -0,0 +1,100 @@
+import java.net.URI;
+import java.util.Optional;
+import java.util.function.Consumer;
+import lombok.AccessLevel;
+import lombok.Getter;
+import org.triplea.http.client.web.socket.messages.ServerMessageEnvelope;
+import org.triplea.http.client.web.socket.messages.WebsocketMessageType;
+
+/**
+ * Used to route server websocket messages to a listener.
+ *
+ * <p>Caution: Do not forget to call the {@code close} method when done listening!
+ *
+ * <p>Very-generic websocket listener! Knows about a set of listeners and message types. The message
+ * types know how to extract a specific message from a server message payload and to send that
+ * message to a listener from the set of listeners. This class wires those together, the net effect
+ * is that after construction a websocket connection is created and when messages are received they
+ * will be routed to the correct listener.
+ *
+ * @param <MessageTypeT> Websocket message type enum.
+ * @param <ListenersTypeT> Listener class, holds a set of consumers, each is an individual consumer.
+ */
+public abstract class WebsocketListener<
+        MessageTypeT extends WebsocketMessageType<ListenersTypeT>, ListenersTypeT>
+    implements Consumer<ServerMessageEnvelope> {
+
+  @Getter(AccessLevel.PROTECTED)
+  private final GenericWebSocketClient webSocketClient;
+
+  private ListenersTypeT listeners;
+
+  private WebsocketListener(
+      final GenericWebSocketClient genericWebSocketClient, final ListenersTypeT listeners) {
+    webSocketClient = genericWebSocketClient;
+    webSocketClient.addMessageListener(this);
+    this.listeners = listeners;
+  }
+
+  /**
+   * Use this constructor if the listener will also be an outbound message sender. Injecting the
+   * websocket client directly allows for easier testing & mocking.
+   */
+  protected WebsocketListener(final GenericWebSocketClient genericWebSocketClient) {
+    this(genericWebSocketClient, null);
+  }
+
+  /**
+   * Constructs a websocket listener that will be connected and messages will be routed to the
+   * listeners passed in.
+   *
+   * @param hostUri Server host (base) uri.
+   * @param websocketPath Path on the server to the websocket to which we will listen.
+   * @param listeners Listener object that contains listeners for each message type we would expect
+   *     to receive from server.
+   */
+  protected WebsocketListener(
+      final URI hostUri, final String websocketPath, final ListenersTypeT listeners) {
+    this(
+        new GenericWebSocketClient(
+            URI.create(hostUri + websocketPath),
+            "Failed to connect to " + URI.create(hostUri + websocketPath)),
+        listeners);
+  }
+
+  public void setListeners(final ListenersTypeT listeners) {
+    this.listeners = listeners;
+  }
+
+  public void close() {
+    webSocketClient.close();
+  }
+
+  @Override
+  public void accept(final ServerMessageEnvelope serverMessageEnvelope) {
+    Preconditions.checkState(listeners != null);
```

Comment:
Question:
Listeners can be set in the constructor or by using the setter.
In which case is the setter required? Because if they can be set in the constructor and aren't modified ever, it would be easy to just add some pre-condition in the constructor to avoid this scenario.

Reply:
If you do not specify the listener in the constructor, then you need to call the setter. This is to support the chat case. I suspect with a moderator lift we could get the chat listeners to be created at the same time as the listeners. I'm not sure how easy that would be as, IIRC, we create the chat websocket essentially on login, we then pass chat to the lobby frame which is the thing that owns and creates the listeners.